eth/stagedsync/stage_polygon_sync.go

```diff
@@ -165,25 +171,214 @@ func SpawnPolygonSyncStage(
 		return err
 	}
 
-	if useExternalTx {
-		return nil
+	if !useExternalTx {
+		return tx.Commit()
+	}
+
+	return nil
+}
+
+func UnwindPolygonSyncStage(ctx context.Context, tx kv.RwTx, u *UnwindState, cfg PolygonSyncStageCfg) error {
+	u.UnwindPoint = max(u.UnwindPoint, cfg.blockReader.FrozenBlocks()) // protect from unwind behind files
+
+	useExternalTx := tx != nil
+	if !useExternalTx {
+		var err error
+		tx, err = cfg.db.BeginRw(ctx)
+		if err != nil {
+			return err
+		}
+
+		defer tx.Rollback()
+	}
+
+	// headers
+	unwindBlock := u.Reason.Block != nil
+	if err := rawdb.TruncateCanonicalHash(tx, u.UnwindPoint+1, unwindBlock); err != nil {
+		return err
+	}
+
+	canonicalHash, err := rawdb.ReadCanonicalHash(tx, u.UnwindPoint)
+	if err != nil {
+		return err
+	}
+
+	if err = rawdb.WriteHeadHeaderHash(tx, canonicalHash); err != nil {
+		return err
+	}
+
+	// bodies
+	if err = cfg.blockWriter.MakeBodiesNonCanonical(tx, u.UnwindPoint+1); err != nil {
+		return err
+	}
+
+	// heimdall
+	if err = UnwindHeimdall(tx, u, nil); err != nil {
+		return err
 	}
 
-	if err := tx.Commit(); err != nil {
+	if err = u.Done(tx); err != nil {
 		return err
 	}
 
+	if !useExternalTx {
+		return tx.Commit()
+	}
+
 	return nil
 }
 
-func UnwindPolygonSyncStage() error {
-	// TODO - headers, bodies (including txnums index), checkpoints, milestones, spans, state sync events
+func UnwindHeimdall(tx kv.RwTx, u *UnwindState, unwindTypes []string) error {
+	if err := UnwindEvents(tx, u, unwindTypes); err != nil {
+		return err
+	}
+
+	if err := UnwindSpans(tx, u, unwindTypes); err != nil {
+		return err
+	}
+
+	if err := UnwindCheckpoints(tx, u, unwindTypes); err != nil {
+		return err
+	}
+
+	if err := UnwindMilestones(tx, u, unwindTypes); err != nil {
+		return err
+	}
+
 	return nil
 }
 
-func PrunePolygonSyncStage() error {
-	// TODO - headers, bodies (including txnums index), checkpoints, milestones, spans, state sync events
-	return nil
+func UnwindEvents(tx kv.RwTx, u *UnwindState, unwindTypes []string) error {
+	if len(unwindTypes) > 0 && !slices.Contains(unwindTypes, "events") {
+		return nil
+	}
+
+	cursor, err := tx.RwCursor(kv.BorEventNums)
+	if err != nil {
+		return err
+	}
+
+	defer cursor.Close()
+	var blockNumBuf [8]byte
+	binary.BigEndian.PutUint64(blockNumBuf[:], u.UnwindPoint+1)
+	k, v, err := cursor.Seek(blockNumBuf[:])
+	if err != nil {
+		return err
+	}
+	if k != nil {
+		// v is the encoding of the first eventId to be removed
```

Comment:
Maybe rename v to eventId, since Seek(v) seems a bit unnatural.

Reply:
`v` stands for `value` as in `kv` (key-value) - does not seem unnatural to me

Reply:
`cursor.Seek(key)` is more typical is what I mean.

Reply:
Can you define "more typical" please?

Reply:
Note this not new code. It is code that I'm moving from 1 place to another.

Reply:
`Seek` takes a key as input . I think maybe I added the comment in the wrong spot.  I meant that it's better to have `k , _ , err := eventCursor.Seek(k)` instead of `v , _ , err := eventCursor.Seek(v)`. So that's why I suggested renaming `v` to `eventId`.

Reply:
Changed to eventId